filebeat/input/filestream/fswatch.go

```diff
@@ -140,140 +119,192 @@ func (w *fileWatcher) watch(ctx unison.Canceler) {
-	return loginp.FSEvent{Op: loginp.OpTruncate, OldPath: path, NewPath: path, Info: fi}
+func truncateEvent(path string, fd loginp.FileDescriptor) loginp.FSEvent {
+	return loginp.FSEvent{Op: loginp.OpTruncate, OldPath: path, NewPath: path, Descriptor: fd}
 }
 
-func renamedEvent(oldPath, path string, fi os.FileInfo) loginp.FSEvent {
-	return loginp.FSEvent{Op: loginp.OpRename, OldPath: oldPath, NewPath: path, Info: fi}
+func renamedEvent(oldPath, path string, fd loginp.FileDescriptor) loginp.FSEvent {
+	return loginp.FSEvent{Op: loginp.OpRename, OldPath: oldPath, NewPath: path, Descriptor: fd}
 }
 
-func deleteEvent(path string, fi os.FileInfo) loginp.FSEvent {
-	return loginp.FSEvent{Op: loginp.OpDelete, OldPath: path, NewPath: "", Info: fi}
+func deleteEvent(path string, fd loginp.FileDescriptor) loginp.FSEvent {
+	return loginp.FSEvent{Op: loginp.OpDelete, OldPath: path, NewPath: "", Descriptor: fd}
 }
 
 func (w *fileWatcher) Event() loginp.FSEvent {
 	return <-w.events
 }
 
-func (w *fileWatcher) GetFiles() map[string]os.FileInfo {
+func (w *fileWatcher) GetFiles() map[string]loginp.FileDescriptor {
 	return w.scanner.GetFiles()
 }
 
+type fingerprintConfig struct {
+	Enabled bool  `config:"enabled"`
+	Offset  int64 `config:"offset"`
+	Length  int64 `config:"length"`
+}
+
 type fileScannerConfig struct {
-	ExcludedFiles []match.Matcher `config:"exclude_files"`
-	IncludedFiles []match.Matcher `config:"include_files"`
-	Symlinks      bool            `config:"symlinks"`
-	RecursiveGlob bool            `config:"recursive_glob"`
+	ExcludedFiles []match.Matcher   `config:"exclude_files"`
+	IncludedFiles []match.Matcher   `config:"include_files"`
+	Symlinks      bool              `config:"symlinks"`
+	RecursiveGlob bool              `config:"recursive_glob"`
+	Fingerprint   fingerprintConfig `config:"fingerprint"`
 }
 
 func defaultFileScannerConfig() fileScannerConfig {
 	return fileScannerConfig{
 		Symlinks:      false,
 		RecursiveGlob: true,
+		Fingerprint: fingerprintConfig{
+			Enabled: false,
+			Offset:  0,
+			Length:  DefaultFingerprintSize,
+		},
 	}
 }
 
-func newFileScanner(paths []string, cfg fileScannerConfig) (loginp.FSScanner, error) {
-	fs := fileScanner{
-		paths:         paths,
-		excludedFiles: cfg.ExcludedFiles,
-		includedFiles: cfg.IncludedFiles,
-		symlinks:      cfg.Symlinks,
-		log:           logp.NewLogger(scannerName),
+// fileScanner looks for files which match the patterns in paths.
+// It is able to exclude files and symlinks.
+type fileScanner struct {
+	paths []string
+	cfg   fileScannerConfig
+	log   *logp.Logger
+}
+
+func newFileScanner(paths []string, config fileScannerConfig) (loginp.FSScanner, error) {
+	s := fileScanner{
+		paths: paths,
+		cfg:   config,
+		log:   logp.NewLogger(scannerDebugKey),
+	}
+
+	if s.cfg.Fingerprint.Enabled {
+		if s.cfg.Fingerprint.Length < sha256.BlockSize {
+			err := fmt.Errorf("fingerprint size %d cannot be smaller than %d", config.Fingerprint.Length, sha256.BlockSize)
```

Comment:
why do we nest a new error twice ? Are we gonna test for the inner one with `error.Is` or similar functions? 
If it's only for formatting a single `fmt.Errorf()` should suffice

Reply:
It's more like a preference to have more structured errors but you're right it's not necessary. Unless there is an argument against it I'd like to keep it.